target/linux/realtek/dts/rtl8380_netgear_gigabit.dtsi

```diff
@@ -0,0 +1,141 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "rtl838x.dtsi"
+
+#include <dt-bindings/input/input.h>
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	compatible = "realtek,rtl838x-soc";
+
+	chosen {
+		bootargs = "console=ttyS0,115200";
+	};
+
+	memory@0 {
+		device_type = "memory";
+		reg = <0x00000000 0x08000000>;
+	};
+
+	keys {
+		compatible = "gpio-keys-polled";
+		poll-interval = <20>;
+
+		mode {
+			label = "reset";
+			gpios = <&gpio0 24 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+	};
+};
+
+&gpio0 {
+	indirect-access-bus-id = <0>;
+};
+
+&spi0 {
+	status = "okay";
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <50000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "u-boot";
+				reg = <0x00000000 0x000e0000>;
+				read-only;
+			};
+
+			partition@e0000 {
+				label = "u-boot-env";
+				reg = <0x000e0000 0x00010000>;
+				read-only;
+			};
+
+			partition@f0000 {
+				label = "sysinfo";
+				reg = <0x000f0000 0x00010000>;
+				read-only;
+			};
+
+			partition@100000{
+				label = "jffs2_cfg";
+				reg = <0x00100000 0x00100000>;
+				read-only;
+			};
+
+			partition@200000{
+				label = "jffs2_log";
+				reg = <0x00200000 0x00100000>;
+				read-only;
+			};
+
+			partition@300000{
+				label = "firmware";
+				compatible = "openwrt,uimage", "denx,uimage";
+				openwrt,ih-magic = <0x4e474520>;
+				openwrt,ih-type = <IH_TYPE_KERNEL>;
+				reg = <0x00300000 0x00e80000>;
+			};
+
+			partition@1180000{
+				label = "firmware2";
+				compatible = "openwrt,uimage", "denx,uimage";
+				openwrt,ih-magic = <0x4e474520>;
+				openwrt,ih-type = <IH_TYPE_KERNEL>;
```

Comment:
Is having these parameters (and compatible) for two different partitions really a good idea?

Reply:
And what is your suggestion here? Those devices are dual boot.

Reply:
And this works OOTB with these parameters now?

Reply:
Yes, it does. You can choose if you would like to boot into OpenWrt or Realtek firmware. By marking it `read-only` for OpenWrt you prevent any accidental modifications of vendor/backup firmware.

Reply:
Well, that's nice, wasn't aware of that.

Reply:
I don't think that's a good idea. There is no dual image support for this target in OpenWrt, so rootfs splitting can only work for one of the partitions.  We should make it explicit which one that is, and so far it has been the first one.

This dts might result in OpenWrt booting from one partition and mounting the rootfs from the other. 

FWIW, I did consider support for using both partitions.  But AFAICS, there is no reliable way to figure out which one we are booting from. The bootpartition variable can be modified in u-boot memeory without writing it back to the sysinfo partition. This might be rare enough to be ignored, if dual partition support was important.  But I don't think it is.  It's just complicating stuff unnecessarily.  So I opted for OpenWrt in the first, and optionally stock in the second image.  Makes everything simple.

This is not unique for this target.  The Unifi AC family has similar separation.

What's unique about these realtek devices is that the stock web application allows directly selecting which partition to flash and boot.  This means that there is no problem reserving one of the partitions for OpenWrt, even if you boot to stock now and then.  You can even upgrade stock without issues.

Reply:
I don't think it is a good idea to make the primary firmware partition the size of the spare flash, as is the case in the current DTS.  Someone might unknowingly overwrite the secondary firmware area.  I see a few possibilities:

1) The partition definition @ 1180000 could just be removed, leaving the partition definition @ 300000 as-is.
2) Keep the second firmware partition as-is, with its read-only attribute.  This could serve as a signal for which partition is booting (the only one that is read-write).
3) Remove the compatible and ih-magic properties and name it something like "reserved".

Any preferences?

Reply:
The partition layout should match what stock firmware has.  I did not know there was a current DTS with something else.

I don't think the rootfs splitting code cares about read-only status, so it might still be confused if both partitions have the openwrt,uimage compatible.  IMHO, it should only be used for partitions we can boot OpenWrt from.

So I prefer 3.  As for the names, I would personally have preferred that we used something closer to stock u-boot and firmware for most of the partitions (except "firmware" obviously, since that is special to sysupgrade).  This is the table from u-boot on my GS108Tv3:
```
RTL838x# flshow
=============== FLASH Partition Layout ===============
Index  Name       Size       Address
------------------------------------------------------
 0     LOADER     0xe0000    0xb4000000-0xb40dffff
 1     BDINFO     0x10000    0xb40e0000-0xb40effff
 2     SYSINFO    0x10000    0xb40f0000-0xb40fffff
 3     JFFS2_CFG  0x100000   0xb4100000-0xb41fffff
 4     JFFS2_LOG  0x100000   0xb4200000-0xb42fffff
 5     RUNTIME1   0xe80000   0xb4300000-0xb517ffff
 6     RUNTIME2   0xe80000   0xb5180000-0xb5ffffff
======================================================
```

The stock firmware also use almost those names when relevant.  Example boot log:

```
Probe: SPI CS1 Flash Type MX25L12845E
Creating 7 MTD partitions on "Total SPI FLASH":
0x00000000-0x00040000 : "LOADER"
0x00040000-0x00050000 : "BDINFO"
0x00050000-0x00060000 : "SYSINFO"
0x00060000-0x00160000 : "JFFS2 CFG"
0x00160000-0x00260000 : "JFFS2 LOG"
0x00260000-0x00930000 : "RUNTIME"
0x00930000-0x01000000 : "RUNTIME2"
```
The differences "RUNTIME" vs "RUNTIME1" and space instead of underscore is still interesting....  But the second image partition is consistently named  "RUNTIME2".  I believe we should use the same name, possibly in lower case.  This will be least confusing to users switching between OpenWrt, u-boot and stock.  

But

Reply:
That all sounds reasonable to me.  I've renamed the partitions in the DTSI to match what the stock firmware uses (aside from the primary runtime partition) and removed the compatible/ih-magic properties from runtime2.